pkg/server/chaosd/kafka.go

```diff
@@ -0,0 +1,353 @@
+}
+
+func getPartitions(attack *core.KafkaCommand) (partitions []int, err error) {
+	dialer, err := newDialer(attack)
+	if err != nil {
+		return nil, err
+	}
+
+	endpoint := fmt.Sprintf("%s:%d", attack.Host, attack.Port)
+	conn, err := dialer.Dial("tcp", endpoint)
+	if err != nil {
+		return nil, errors.Wrapf(err, "dial endpoint: %s", endpoint)
+	}
+
+	if attack.Topic == "" {
+		return nil, errors.New("topic is required")
+	}
+	pars, err := conn.ReadPartitions(attack.Topic)
+	if err != nil {
+		return nil, errors.Wrapf(err, "read partitions of topic %s", attack.Topic)
+	}
+	for _, par := range pars {
+		if par.Error != nil {
+			return nil, errors.Wrap(err, "read partition")
+		}
+		partitions = append(partitions, par.ID)
+	}
+	return partitions, nil
+}
+
+func attackKafkaFill(ctx context.Context, attack *core.KafkaCommand) (err error) {
+	// TODO: make it configurable
+	const messagePerRequest = 128
+	msg := make([]byte, attack.MessageSize)
+	msgList := make([]client.Message, 0, messagePerRequest)
+	for i := 0; i < messagePerRequest; i++ {
+		msgList = append(msgList, client.Message{Topic: attack.Topic, Value: msg})
+	}
+
+	partitions, err := getPartitions(attack)
+	if err != nil {
+		return err
+	}
+
+	writeChan := make(chan uint64)
+	errChan := make(chan error)
+	c, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for _, partition := range partitions {
+		p := partition
+		go func() {
+			conn, err := dialPartition(c, attack, p)
+			if err != nil {
+				errChan <- err
+				return
+			}
+
+			for c.Err() == nil {
+				n, err := conn.WriteMessages(msgList...)
+				if err != nil {
+					errChan <- perr.Wrap(err, "write messages")
+					return
+				}
+				writeChan <- uint64(n)
+			}
+		}()
+	}
+
+	start := time.Now()
+	written := "0 B"
+	bytes := uint64(0)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-errChan:
+			return err
+		case n := <-writeChan:
+			bytes += uint64(n)
+			newWritten := humanize.Bytes(bytes)
+			if newWritten != written {
+				written = newWritten
+				log.Info(fmt.Sprintf("write %s in %s", written, time.Now().Sub(start)))
+			}
+			if bytes >= attack.MaxBytes {
+				return nil
+			}
+		}
+	}
+}
+
+func attackKafkaFlood(ctx context.Context, attack *core.KafkaCommand) (err error) {
+	msg := make([]byte, attack.MessageSize)
+	wg := new(sync.WaitGroup)
+	for i := 0; i < int(attack.Threads); i++ {
+		logger := log.With(zap.String("thread", fmt.Sprintf("thread-%d", i)))
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			conn, err := dialPartition(ctx, attack, 0)
+			if err != nil {
+				logger.Error(perr.Wrapf(err, "dial kafka broker: %s", fmt.Sprintf("%s:%d", attack.Host, attack.Port)).Error())
+				return
+			}
+			defer conn.Close()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					start := time.Now()
+					succeeded := 0
+					failed := 0
+					counter := uint64(0)
+					for time.Now().Sub(start) < time.Second && counter < attack.RequestPerSecond {
```

Comment:
In my opinion, we can remove the config `RequestPerSecond`, chaosd can write messages as fast as possible, and then we can remove this complex logic below.

Reply:
> Every time get n from channel writeChan means had already written some message to Kafka, I think it's not necessary to add this judgment, print the log below even if chaosd write same size message with last time

Because we use the humanize to format bytes, there may be the same `written` in many lines when the written is too large.